### DISCUSSION TITLE
Pixel count on canvas not correct

### DISCUSSION THREAD

**Discussion by DanielBruijn** [jspsych/jsPsych#1764] [Q&A] [answered]
Hello all, 

In my experiment i display canvases that need to be as big as my screen, 1080x1920 pixels. However, when i use
```canvas_size: [1080,1920]```
i keep seeing scrollbars on the side and bottom, indicating that the canvas is bigger than my screen. I decreased gradually to 
```canvas_size: [1073,1872]``` 
and then it stops showing scrollbars. How is this possible? It's a small difference, but big enough to be annoying. Also, i have a fixation cross in the middle of the canvas, which is no longer in the middle, but a bit to the right of the middle. 

Can anyone explain/suggest how to fix this? For my experiment, the size of objects is rather important, so i need to know for sure that 66 pixels equal 1 cm, for instance.

Is there maybe a way to do this?:
```canvas_size: [participant_user_resolution]```

I'm open to any suggestion, thanks in advance!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @DanielBruijn, if you open the browser's developer tools and find the Elements tab, you'll see that the canvas element is actually nested under a few different things: body -> jspsych-content-wrapper-div -> jspsych-content div -> jspsych-canvas-keyboard-response-stimulus div -> canvas element. Most of these things have a height and width of 100%, except for the jspsych-content div, which has a max-width of 95%. This means that, by default, all of the jsPsych content will be slightly more narrow than the page width. You can change that like this:
```html
<style>#jspsych-content {max-width:100%;}</style>
```
Then, to get the size of the browser page to use for the canvas height and width, you can use the `clientHeight` and `clientWidth` of the `<body>` element.
```js
var canvas_h = document.getElementsByTagName('body')[0].clientHeight;
var canvas_w = document.getElementsByTagName('body')[0].clientWidth;
```
I hope this helps. Let us know if it works, and if you have any other questions.

**Reply by DanielBruijn** [reply]
Hey Becky, thank you for your suggestion! I work in cognition.run, so i normally don't use style elements like that.  I have an adjusted the CSS file (see below), which sets that max-width parameter to 100%. I assume that gets that part of the job done?

```
/*
 * CSS for jsPsych experiments.
 *
 * This stylesheet provides minimal styling to make jsPsych
 * experiments look polished without any additional styles.
 */

(...)

 .jspsych-content {
   **max-width: 100%**; /* this is mainly an IE 10-11 fix */
   text-align: center;
   margin: auto; /* this is for overflowing content */
 }

```
Then for part two, using the `clientHeight` and `clientWidth` i get the same 1073 by 1872 size.

To be honest, i don't think this is the issue, since if it were indeed 95% scaling, it should size to 1026 by 1824, not to the specific sizes i encounter. Any other ideas?

Thanks so far!

**Reply by becky-gilbert** [reply]
Ah sorry, yes if you're using cognition.run then editing the jsPsych CSS file (or uploading a separate one) is the way to go!

> Then for part two, using the clientHeight and clientWidth i get the same 1073 by 1872 size.

What happens when you use those values for the canvas size with the jspsych-content max-width CSS change?

> To be honest, i don't think this is the issue, since if it were indeed 95% scaling, it should size to 1026 by 1824, not to the specific sizes i encounter. Any other ideas?

When the jspsych-content div is restricted to 95% width, then the result would not be that the canvas is scaled down to fit inside of it - rather, the canvas will still be the fixed size that you gave it, but the left side of the canvas will start at the edge of the jspsych-content div (rather than the left edge of the page), which will push the canvas to the right a bit, so it will extend past the right side of the page and you'll get scroll bars. So I think that jspsych-content div's max-width is part of the problem. But maybe there's another reason the canvas doesn't fit on the page... Feel free to share a link to your experiment and I can have a look.

**Reply by DanielBruijn** [reply]
> What happens when you use those values for the canvas size with the jspsych-content max-width CSS change?

I already used the combination of CSS and [canvas_h,canvas_w], if that is what you meant.



> but the left side of the canvas will start at the edge of the jspsych-content div (rather than the left edge of the page), which will push the canvas to the right a bit, so it will extend past the right side of the page and you'll get scroll bars.

I see, that is indeed what happens when i input [1080,1920] as the size.

The link to my experiment is 33rnlr23rk.cognition.run. Feel free to nose around! It's not very urgent though, my experiment is already live with the slightly smaller canvas. I also use the `jsPsych-resize` plugin, so the size is a little bit different for everybody which is fine as long as the deviations are not excessive.

**Reply by becky-gilbert** [reply]
> I already used the combination of CSS and [canvas_h,canvas_w], if that is what you meant.

And you're still getting scroll bars? 

I tried checking myself but I wasn't seeing the canvas on the page, and when I looked at the Elements tab, I saw that the canvas was on the page but had a height of 0. I think the reason might be because the body element isn't ready yet when you get its `clientHeight`. You could try getting the height and width values at some point during the experiment, before the canvas trials start. But then you'd also have to make the `canvas_size` parameter into a function so that you get the updated values for the two variables, e.g.
```js
var canvas_h;
var canvas_w;

practice_instructions = {
  type: "html-keyboard-response",
  stimulus: [
    "<p style='font-size: 25px'> First we'll do a practice round, with easier trials (the first screen will appear for a longer time).</p>" + 
    "<p style='font-size: 25px'> Remember, press 'C' if the indicated circle has Changed colour, and press 'N' if it has Not changed colour. </p>" +
    "<p style='color:#D3D3D3'>This text is just to create space.</p>" +
    "<p style='font-size: 25px'> Press 'P' to proceed and have fun!</p>"
    ],
  choices: "p",
  on_finish: function() {
    canvas_h = document.getElementsByTagName('body')[0].clientHeight;
    canvas_w = document.getElementsByTagName('body')[0].clientWidth;
  }
};

var practice_trials = {
  timeline: [
    {
      type: 'canvas-keyboard-response',
      stimulus: drawBlank,
      choices: jsPsych.NO_KEYS,
      trial_duration: 100-8, //the -8 represents the mean few milliseconds of lag that the browser can introduce. Rather a shorter presentation then a longer one, especially for the delays/SOAs
      canvas_size: function() { return [canvas_h,canvas_w]; }
    },
  // ...
```
Also, you could try hiding the scroll bars like this:
```js
document.body.style.overflow = "hidden";
```

**Reply by DanielBruijn** [reply]
> And you're still getting scroll bars?

No, because [1073,1872] is the size at which the scrollbars disappear, and apparently canvas_h/w take on the same values.


> You could try getting the height and width values at some point during the experiment,

Right, i will try that. 

> Also, you could try hiding the scroll bars like this:
> 
> ```js
> document.body.style.overflow = "hidden";
> ```

Indeed, i was using that already ;)

**Reply by DanielBruijn** [reply]
Oh okay, that actually worked! First it didn't, but after a reload it did, hence my surprise. So now i have defined the `canvas_size` as functions that retrieve `clientHeight` and `clientWidth`, and that works. I checked that it actually presented a [1080,1920] canvas by putting the fixation cross at [960,540] (instead of defining it as being halfway the canvas width and height), and it turned out to be right in the middle. 

So, problem solved, thanks! Odd though, that it takes this much effort to get a few pixels right.

**Reply by becky-gilbert** [reply]
Great! 

> Odd though, that it takes this much effort to get a few pixels right.

I agree, sorry... Maybe there's an easier way to do this that I'm not aware of.
